File: app/src/main/java/com/nurram/project/imagetextrecognition/MainActivity.java
package com.nurram.project.imagetextrecognition;

import android.content.Intent;
import android.graphics.Bitmap;
import android.media.MediaPlayer;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.FloatingActionButton;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.View;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.RelativeLayout;
import android.widget.Toast;

import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.firebase.ml.vision.FirebaseVision;
import com.google.firebase.ml.vision.common.FirebaseVisionImage;
import com.google.firebase.ml.vision.text.FirebaseVisionText;
import com.google.firebase.ml.vision.text.FirebaseVisionTextRecognizer;
import com.wonderkiln.camerakit.CameraKitError;
import com.wonderkiln.camerakit.CameraKitEvent;
import com.wonderkiln.camerakit.CameraKitEventListener;
import com.wonderkiln.camerakit.CameraKitImage;
import com.wonderkiln.camerakit.CameraKitVideo;
import com.wonderkiln.camerakit.CameraView;

import java.util.ArrayList;

public class MainActivity extends AppCompatActivity {
    private CameraView mCameraView;
    private ImageButton mCameraButton;
    private ProgressBar mCameraProgress;
    private ImageView mImageView;
    private RelativeLayout mCapturedLayout;
    private FloatingActionButton mProcessImage;

    private boolean isCapturedShow = false;
    private FirebaseVisionImage mImage;
    private ArrayList<String> mBlockList;
    private FirebaseVisionTextRecognizer mRecognizer;
    private Intent mIntent;
    private MediaPlayer mPlayer;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        mCameraProgress = findViewById(R.id.progress);
        mCapturedLayout = findViewById(R.id.captured_image);
        mImageView = findViewById(R.id.image);
        mBlockList = new ArrayList<>();

        mCameraView = findViewById(R.id.camera_layout);
        mCameraView.addCameraKitListener(new CameraKitEventListener() {
            @Override
            public void onEvent(CameraKitEvent cameraKitEvent) {}

            @Override
            public void onError(CameraKitError cameraKitError) {}

            @Override
            public void onImage(CameraKitImage cameraKitImage) {
                Bitmap bitmap = cameraKitImage.getBitmap();
                bitmap = Bitmap.createBitmap(bitmap);

                mImage = FirebaseVisionImage.fromBitmap(bitmap);

                mCapturedLayout.setVisibility(View.VISIBLE);
                mCameraProgress.setVisibility(View.GONE);
                isCapturedShow = true;

                mImageView.setImageBitmap(mImage.getBitmapForDebugging());
            }

            @Override
            public void onVideo(CameraKitVideo cameraKitVideo) {}
        });

        mCameraButton = findViewById(R.id.button_camera);
        mCameraButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Toast.makeText(MainActivity.this, "Tahan sampai proses selesai", Toast.LENGTH_SHORT).show();
                mCameraButton.setVisibility(View.GONE);
                mCameraProgress.setVisibility(View.VISIBLE);

                mCameraView.captureImage();

                mPlayer = MediaPlayer.create(MainActivity.this, R.raw.camera);
                mPlayer.start();

            }
        });

        mProcessImage = findViewById(R.id.process_fab);
        mProcessImage.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                mProcessImage.setClickable(false);
                processTextRecog(mImage);
            }
        });

    }

    @Override
    protected void onResume() {
        super.onResume();
        mCameraView.start();
        mCameraButton.setVisibility(View.VISIBLE);
        mCameraProgress.setVisibility(View.GONE);
        mProcessImage.setClickable(true);
    }

    @Override
    protected void onStop() {
        super.onStop();
        mCameraView.stop();

        if (mPlayer != null) {
            mPlayer.stop();
        }

    }

    @Override
    public void onBackPressed() {
        if (isCapturedShow) {
            mCapturedLayout.setVisibility(View.GONE);
            mCameraProgress.setVisibility(View.GONE);
            mCameraButton.setVisibility(View.VISIBLE);
            isCapturedShow = false;
        } else {
            super.onBackPressed();
        }
    }

    private void processTextRecog(FirebaseVisionImage image) {
        mBlockList.clear();

        mRecognizer = FirebaseVision.getInstance()
                .getOnDeviceTextRecognizer();

        mRecognizer.processImage(image)
                .addOnSuccessListener(new OnSuccessListener<FirebaseVisionText>() {
                    @Override
                    public void onSuccess(FirebaseVisionText result) {
                        mIntent = ResultActivity.getIntent(MainActivity.this);
                        mCapturedLayout.setVisibility(View.GONE);

                        for (FirebaseVisionText.TextBlock block : result.getTextBlocks()) {
                            for (FirebaseVisionText.Line line : block.getLines()) {
                                String lineText = line.getText();
                                mBlockList.add(lineText);
                            }
                        }

                        mIntent.putStringArrayListExtra("block", mBlockList);

                        startActivity(mIntent);
                    }
                })
                .addOnFailureListener(
                        new OnFailureListener() {
                            @Override
                            public void onFailure(@NonNull Exception e) {
                                Log.e("TAG", e.getMessage());
                            }
                        });
    }
}

File: app/src/main/java/com/nurram/project/imagetextrecognition/MainActivity.kt
package com.nurram.project.imagetextrecognition

import android.content.Intent
import android.graphics.Bitmap
import android.net.Uri
import android.os.Bundle
import android.os.Environment
import android.provider.MediaStore
import android.view.View
import android.widget.*
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.FileProvider
import androidx.lifecycle.lifecycleScope
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.text.Text
import com.google.mlkit.vision.text.TextRecognition
import com.google.mlkit.vision.text.latin.TextRecognizerOptions
import com.nurram.project.imagetextrecognition.databinding.ActivityMainBinding
import com.squareup.picasso.Picasso
import kotlinx.coroutines.DelicateCoroutinesApi
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.*
import java.text.SimpleDateFormat
import java.util.*


@DelicateCoroutinesApi
class MainActivity : AppCompatActivity() {
    private lateinit var binding: ActivityMainBinding

    private var mBlockList = arrayListOf<String>()
    private lateinit var currentPhotoPath: String

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        Intent(MediaStore.ACTION_IMAGE_CAPTURE).also { takePictureIntent ->
            // Ensure that there's a camera activity to handle the intent
            takePictureIntent.resolveActivity(packageManager)?.also {
                // Create the File where the photo should go
                val photoFile: File? = try {
                    createImageFile()
                } catch (ex: IOException) {
                    ex.printStackTrace()
                    null
                }
                // Continue only if the File was successfully created
                photoFile?.also {
                    val photoURI: Uri = FileProvider.getUriForFile(
                        this,
                        "com.example.android.fileprovider",
                        it
                    )
                    takePictureIntent.putExtra(MediaStore.EXTRA_OUTPUT, photoURI)
                    startActivityForResult(takePictureIntent, INTENT_CAMERA)
                }
            }
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == INTENT_CAMERA && resultCode == RESULT_OK) {

            lifecycleScope.launch {
                var bitmap: Bitmap

                binding.mainProgress.visibility = View.VISIBLE
                withContext(lifecycleScope.coroutineContext + Dispatchers.IO) {
                    bitmap = Picasso.get().load("file:$currentPhotoPath").get()
                }
                binding.mainProgress.visibility = View.GONE
                binding.capturedLayout.capturedImage.visibility = View.VISIBLE

                binding.capturedLayout.image.setImageBitmap(bitmap)
                binding.capturedLayout.processFab.setOnClickListener {
                    processTextRecog(bitmap)
                }
            }
        }
    }


    override fun onResume() {
        super.onResume()
        binding.capturedLayout.processFab.isClickable = true
    }

    @Throws(IOException::class)
    private fun createImageFile(): File {
        val timeStamp: String = SimpleDateFormat("yyyyMMdd_HHmmss").format(Date())
        val storageDir: File? = getExternalFilesDir(Environment.DIRECTORY_PICTURES)
        return File.createTempFile(
            "JPEG_${timeStamp}_",
            ".jpg",
            storageDir
        ).apply { currentPhotoPath = absolutePath }
    }

    private fun processTextRecog(image: Bitmap) {
        val inputImage = InputImage.fromBitmap(image, 0)
        mBlockList.clear()

        val mRecognizer = TextRecognition.getClient(TextRecognizerOptions.DEFAULT_OPTIONS)
        mRecognizer.process(inputImage)
            .addOnSuccessListener { result: Text ->

                for (block in result.textBlocks) {
                    for (line in block.lines) {
                        val lineText = line.text
                        mBlockList.add(lineText)
                    }
                }

                val intent = ResultActivity.getIntent(this@MainActivity)
                intent.putStringArrayListExtra("block", mBlockList)
                startActivity(intent)

                finish()
            }
            .addOnFailureListener {
                Toast.makeText(
                    this,
                    "Something happened",
                    Toast.LENGTH_SHORT
                ).show()
            }
    }

    companion object {
        const val INTENT_CAMERA = 101
    }
}